### PR TITLE
RTMDet Assigner visualization

### PR DESCRIPTION
**Pull request by yechenzhi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation for this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a potential influence on downstream projects, this PR should be tested with downstream projects, like MMDetection or MMClassification.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmyolo/models/task_modules/assigners/batch_dsl_assigner.py [hunk 1]
```diff
@@ -146,7 +146,8 @@ def forward(self, pred_bboxes: Tensor, pred_scores: Tensor, priors: Tensor,
             assigned_labels=assigned_labels,
             assigned_labels_weights=assigned_labels_weights,
             assigned_bboxes=assigned_bboxes,
-            assign_metrics=assign_metrics)
+            assign_metrics=assign_metrics,
+            matched_gt_inds=matched_gt_inds)
```
Comment (hhaAndroid) [inline mmyolo/models/task_modules/assigners/batch_dsl_assigner.py:150]: Why do we need to change?

File: projects/assigner_visualization/assigner_visualization.py [hunk 2]
```diff
@@ -82,11 +86,17 @@ def main():
 
     # build model
     model = MODELS.build(cfg.model)
-    assert isinstance(model.bbox_head, YOLOv5HeadAssigner),\
-        'Now, this script only support yolov5, and bbox_head must use ' \
-        '`YOLOv5HeadAssigner`. Please use `' \
+    if args.checkpoint is not None:
+        checkpoint = _load_checkpoint(args.checkpoint)
```
Comment (hhaAndroid) [inline projects/assigner_visualization/assigner_visualization.py:90]: load checkpoint to model 不需要调用两个api，有一个对外的api 一行就可以

File: projects/assigner_visualization/assigner_visualization.py [hunk 3]
```diff
@@ -82,11 +86,17 @@ def main():
 
     # build model
     model = MODELS.build(cfg.model)
-    assert isinstance(model.bbox_head, YOLOv5HeadAssigner),\
-        'Now, this script only support yolov5, and bbox_head must use ' \
-        '`YOLOv5HeadAssigner`. Please use `' \
+    if args.checkpoint is not None:
```
Comment (Nioolek) [inline projects/assigner_visualization/assigner_visualization.py:89]: 这里打印一下，如果是rtmdethead，建议load_checkpoint

Comment (Nioolek) [reply projects/assigner_visualization/assigner_visualization.py:89]: 我在想是不是可以默认调用在coco上训练好的模型

File: projects/assigner_visualization/detectors/rtm_detector_assigner.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Union
+
+from mmyolo.models import YOLODetector
+from mmyolo.registry import MODELS
+
+
+@MODELS.register_module()
+class RTMDetectorAssigner(YOLODetector):
```
Comment (Nioolek) [inline projects/assigner_visualization/detectors/rtm_detector_assigner.py:9]: 好不好直接改YOLODetectorAssigner，用参数来控制是否要推理backbone和neck

File: projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                Samples. It usually includes information such as
+                `gt_instance`, `gt_panoptic_seg` and `gt_sem_seg`.
+            inputs_hw: Height and width of inputs size
+
+        Returns:
+            dict: A dictionary of assigning components.
+        """
+        if isinstance(x, list):
+            # TODO: fix here, not implemented for normal version
+            pass
```
Comment (Nioolek) [inline projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py:168]: 没实现的话就直接`raise NotImplentError`

File: projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    def assign(self, x: List[Tensor], batch_gt_instances: InstanceList,
+               batch_img_metas: List[dict]) -> dict:
+        """Calculate assigning results. This function is provided to the
+        `assigner_visualization.py` script.
+
+        Args:
+            batch_data_samples (List[:obj:`DetDataSample`], dict): The Data
+                Samples. It usually includes information such as
+                `gt_instance`, `gt_panoptic_seg` and `gt_sem_seg`.
+            inputs_hw: Height and width of inputs size
```
Comment (Nioolek) [inline projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py:161]: del this

File: mmyolo/models/task_modules/assigners/batch_dsl_assigner.py (hunk 1 above)
Comment (yechenzhi) [reply mmyolo/models/task_modules/assigners/batch_dsl_assigner.py:150]: we need matched_gt_inds to find the labeled bounding box corresponding to the positive samples. Do you know a simpler implementation？

Comment (yechenzhi) [reply mmyolo/models/task_modules/assigners/batch_dsl_assigner.py:150]: I guess I can use assigned_bboxes to find the corresponding labeled bounding box, I'll try it later.

File: projects/assigner_visualization/detectors/rtm_detector_assigner.py (hunk 4 above)
Comment (yechenzhi) [reply projects/assigner_visualization/detectors/rtm_detector_assigner.py:9]: I'll try it later

File: projects/assigner_visualization/README.md [hunk 7]
```diff
@@ -15,3 +15,9 @@ Now, the script only support `YOLOv5` .
 ```shell
 python projects/assigner_visualization/assigner_visualization.py projects/assigner_visualization/configs/yolov5_s-v61_syncbn_fast_8xb16-300e_coco_assignervisualization.py
 ```
+
+```shell
+python projects/assigner_visualization/assigner_visualization.py projects/assigner_visualization/configs/rtmdet_s_syncbn_fast_8xb32-300e_coco_assignervisualization.py -c ${checkpont}
+```
+
+${checkpont} is the checkpont file path, if there is no such file, the positive samples of RTMDet will be given by the random initialization model.
```
Comment (Nioolek) [inline projects/assigner_visualization/README.md:23]: 加下文档：RTMDet采用动态assigner，模型权重会影响样本分配结果，所以推荐加载模型使用

File: projects/assigner_visualization/configs/rtmdet_s_syncbn_fast_8xb32-300e_coco_assignervisualization.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+_base_ = ['../../../configs/rtmdet/rtmdet_s_syncbn_fast_8xb32-300e_coco.py']
+
+custom_imports = dict(imports=[
+    'projects.assigner_visualization.detectors',
+    'projects.assigner_visualization.dense_heads'
+])
+
+model = dict(
+    type='YOLODetectorAssigner', bbox_head=dict(type='RTMHeadAssigner'))
```
Comment (Nioolek) [inline projects/assigner_visualization/configs/rtmdet_s_syncbn_fast_8xb32-300e_coco_assignervisualization.py:9]: 这里加load_from='http://' 已经训练好的模型，能不能实现不初始化权重，而采用已经训练好的权重？这样会不会更好？ @hhaAndroid

File: projects/assigner_visualization/README.md [hunk 9]
```diff
@@ -15,3 +15,9 @@ Now, the script only support `YOLOv5` .
 ```shell
 python projects/assigner_visualization/assigner_visualization.py projects/assigner_visualization/configs/yolov5_s-v61_syncbn_fast_8xb16-300e_coco_assignervisualization.py
```
Comment (Nioolek) [inline projects/assigner_visualization/README.md:16]: 标注一下，yolov5不需要加载checkpoint

File: projects/assigner_visualization/README.md [hunk 10]
```diff
@@ -15,3 +15,9 @@ Now, the script only support `YOLOv5` .
...
 ```
+
+```shell
+python projects/assigner_visualization/assigner_visualization.py projects/assigner_visualization/configs/rtmdet_s_syncbn_fast_8xb32-300e_coco_assignervisualization.py -c ${checkpont}
+```
+
```
Comment (Nioolek) [inline projects/assigner_visualization/README.md:22]: 加个文档链接跳转：如果想了解assigner细节，可以查看文档

File: projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                grid_y_inds = retained_pos_inds // w
+                grid_x_inds = retained_pos_inds - retained_pos_inds // w * w
+                assign_results_prior = {
+                    'stride': self.featmap_strides[i],
+                    'grid_x_inds': grid_x_inds,
+                    'grid_y_inds': grid_y_inds,
+                    'img_inds': img_inds[retained_inds],
+                    'class_inds': labels[retained_inds],
+                    'retained_gt_inds': matched_gt_inds[retained_inds],
+                    'prior_ind': 1
```
Comment (Nioolek) [inline projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py:139]: `0` or `1`?

Comment (yechenzhi) [reply projects/assigner_visualization/dense_heads/rtmdet_head_assigner.py:139]: 0，my bad

**Review by Nioolek — APPROVED** [review]
LGTM

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**